File: src/main/java/com/mparticle/kits/TaplyticsKit.java
package com.mparticle.kits;

import android.content.Context;

import org.json.JSONException;

import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.Collections;
import java.math.BigDecimal;

import com.mparticle.MPEvent;
import com.mparticle.MParticle;
import com.mparticle.commerce.CommerceEvent;
import com.mparticle.commerce.Product;
import com.mparticle.commerce.TransactionAttributes;

import com.mparticle.identity.MParticleUser;
import com.taplytics.sdk.Taplytics;
import com.taplytics.sdk.TaplyticsHasUserOptedOutListener;
import com.taplytics.sdk.TaplyticsResetUserListener;

import org.json.JSONObject;

import static com.mparticle.MParticle.IdentityType;

public class TaplyticsKit extends KitIntegration
        implements
        KitIntegration.AttributeListener,
        KitIntegration.EventListener,
        KitIntegration.CommerceListener,
        KitIntegration.IdentityListener {

    /**
     * Option Keys
     */
    private static final String API_KEY = "apiKey";
    private static final String AGGRESSIVE = "TaplyticsOptionAggressive";
    private static final String TAPLYTICS_AGGRESSIVE = "aggressive";
    private static final String USER_ID = "user_id";
    private static final String EMAIL = "email";
    private static final String DELAYED_START = "delayedStartTaplytics";

    /**
     * tlOptions get and set methods
     */

    private static Map<String, Object> tlOptions = new HashMap<>();

    public static Map<String, Object> getTlOptions() {
        return tlOptions;
    }

    public static void setTlOptions(Map<String, Object> options) {
        tlOptions = options;
    }

    private HashMap<String, Object> mergeOptions(Map<String, Object> tlOptions, Map<String, Object> configuration) {
        if (tlOptions == null) {
            tlOptions = new HashMap<>();
        }
        if (configuration == null) {
            configuration = new HashMap<>();
        }
        HashMap<String, Object> merged = new HashMap<>(configuration);
        for (Map.Entry<String, Object> entry : tlOptions.entrySet()) {
            merged.put(entry.getKey(), entry.getValue());
        }
        return merged;
    }

    @Override
    protected List<ReportingMessage> onKitCreate(Map<String, String> settings, Context context) {
        startTaplytics(settings, context);
        return null;
    }

    private List<ReportingMessage> createReportingMessages(String message) {
        ReportingMessage reportingMessage = new ReportingMessage(this,
                message,
                System.currentTimeMillis(),
                null);
        return Collections.singletonList(reportingMessage);
    }

    private void startTaplytics(Map<String, String> settings, Context context) {
        String apiKey = getAPIKey(settings);
        HashMap<String, Object> options = mergeOptions(getTlOptions(), getOptionsFromConfiguration(settings));
        options.put(DELAYED_START, true);
        Taplytics.startTaplytics(context, apiKey, options);
    }

    private String getAPIKey(Map<String, String> settings) {
        final String apiKey = settings.get(API_KEY);
        if (KitUtils.isEmpty(apiKey)) {
            throw new IllegalArgumentException("Failed to initialize Taplytics SDK - an API key is required");
        }
        return apiKey;
    }


    private Map<String, Object> getOptionsFromConfiguration(Map<String, String> settings) {
        Map<String, Object> options = new HashMap<>();
        addAggressiveOption(options, settings);

        return options.isEmpty() ? null : options;
    }

    private void addAggressiveOption(Map<String, Object> options, Map<String, String> settings) {
        Boolean agg = Boolean.parseBoolean(settings.get(AGGRESSIVE));
        options.put(TAPLYTICS_AGGRESSIVE, agg.booleanValue());
    }

    @Override
    public String getName() {
        return "Taplytics";
    }

    /**
     * AttributeListener Interface
     */

    @Override
    public void setUserAttribute(String attributeKey, String attributeValue) {
        try {
            JSONObject attr = new JSONObject();
            if (attributeValue != null) {
                attr.put(attributeKey, attributeValue);
            } else {
                attr.put(attributeKey, "");
            }
            Taplytics.setUserAttributes(attr);
        } catch (JSONException e) {

        }
    }

    @Override
    public boolean supportsAttributeLists() { return false; }

    @Override
    public void setAllUserAttributes(Map<String, String> attributes, Map<String, List<String>> attributeLists) {
        for (Map.Entry<String, String> entry : attributes.entrySet()) {
            setUserAttribute(entry.getKey(), entry.getValue());
        }
    }

    @Override
    public void setUserIdentity(MParticle.IdentityType identityType, String identity) {
        switch (identityType) {
            case CustomerId: {
                setUserAttribute(USER_ID, identity);
                break;
            }
            case Email: {
                setUserAttribute(EMAIL, identity);
            }
        }
    }

    @Override
    public void removeUserIdentity(MParticle.IdentityType identityType) {
        setUserIdentity(identityType, null);
    }


    @Override
    public void removeUserAttribute(String attribute) {
        setUserAttribute(attribute, null);
    }

    @Override
    public void onIdentifyCompleted(MParticleUser mParticleUser, FilteredIdentityApiRequest filteredIdentityApiRequest) {
        setUserAttributeFromRequest(filteredIdentityApiRequest);
    }

    /**
     * Identity Listener
     */

    public void setUserAttributeFromRequest(FilteredIdentityApiRequest request) {
        Map<MParticle.IdentityType, String> identities = request.userIdentities;
        try {
            JSONObject attr = new JSONObject();
            if (identities.get(IdentityType.CustomerId) != null) {
                attr.put(USER_ID, identities.get(IdentityType.CustomerId));
            }
            if (identities.get(IdentityType.Email) != null) {
                attr.put(EMAIL, identities.get(IdentityType.Email));
            }
            Taplytics.setUserAttributes(attr);
        } catch (JSONException e) {

        }
    }

    @Override
    public void onLoginCompleted(MParticleUser user, FilteredIdentityApiRequest request) {
        setUserAttributeFromRequest(request);
    }

    @Override
    public void onLogoutCompleted(MParticleUser mParticleUser, FilteredIdentityApiRequest filteredIdentityApiRequest) {
        Taplytics.resetAppUser(new TaplyticsResetUserListener() {
            @Override
            public void finishedResettingUser() {
                // no-op
            }
        });
    }

    @Override
    public void onModifyCompleted(MParticleUser mParticleUser, FilteredIdentityApiRequest filteredIdentityApiRequest) {
        // no-op
    }

    @Override
    public void onUserIdentified(MParticleUser mParticleUser) {
        // no-op
    }


    /**
     * Unsupported methods
     */
    @Override
    public List<ReportingMessage> logout() { return null; }

    @Override
    public void setUserAttributeList(String attribute, List<String> attributeValueList) { }

    /**
     * CommerceListener Interface
     */

    @Override
    public List<ReportingMessage> logEvent(CommerceEvent event) {
        if (!KitUtils.isEmpty(event.getProductAction()) &&
                event.getProductAction().equalsIgnoreCase(Product.PURCHASE)) {

            TransactionAttributes transactionAttributes = event.getTransactionAttributes();

            if (transactionAttributes == null) {
                return null;
            }

            String id = transactionAttributes.getId();
            Double revenue = transactionAttributes.getRevenue();

            if (id == null || revenue == null) {
                return null;
            }

            Taplytics.logRevenue(id, revenue);
            return Collections.singletonList(ReportingMessage.fromEvent(this, event));
        }
        return null;
    }

    /**
     * EventListener Interface
     */

    @Override
    public List<ReportingMessage> logEvent(MPEvent event) {
        String eventName = event.getEventName();
        Taplytics.logEvent(eventName);
        return Collections.singletonList(ReportingMessage.fromEvent(this, event));
    }

    @Override
    public List<ReportingMessage> logScreen(String screenName, Map<String, String> screenAttributes) {
        Taplytics.logEvent(screenName);
        return createReportingMessages(ReportingMessage.MessageType.SCREEN_VIEW);
    }

    /**
     * Unsupported Methods
     */

    @Override
    public List<ReportingMessage> logException(Exception exception, Map<String, String> exceptionAttributes, String message) { return null; }

    @Override
    public List<ReportingMessage> logError(String message, Map<String, String> errorAttributes) { return null; }

    @Override
    public List<ReportingMessage> leaveBreadcrumb(String breadcrumb) { return null; }

    //put all these together
    @Override
    public List<ReportingMessage> logLtvIncrease(BigDecimal valueIncreased, BigDecimal valueTotal, String eventName, Map<String, String> contextInfo) { return null; }

    @Override
    public List<ReportingMessage> setOptOut(final boolean optedOut) {
        Taplytics.hasUserOptedOutTracking(null, new TaplyticsHasUserOptedOutListener() {
            @Override
            public void hasUserOptedOutTracking(boolean hasOptedOut) {
                if (!hasOptedOut && optedOut) {
                    Taplytics.optOutUserTracking(null);
                } else if (hasOptedOut && !optedOut) {
                    Taplytics.optInUserTracking(null);
                }
            }
        });

        return createReportingMessages(ReportingMessage.MessageType.OPT_OUT);
    }
}

File: src/main/kotlin/com/mparticle/kits/TaplyticsKit.kt
package com.mparticle.kits

import android.content.Context
import com.mparticle.MPEvent
import com.mparticle.MParticle.IdentityType
import com.mparticle.commerce.CommerceEvent
import com.mparticle.commerce.Product
import com.mparticle.identity.MParticleUser
import com.mparticle.kits.KitIntegration.AttributeListener
import com.mparticle.kits.KitIntegration.CommerceListener
import com.mparticle.kits.KitIntegration.IdentityListener
import com.mparticle.kits.KitIntegration.SessionListener
import com.taplytics.sdk.Taplytics
import org.json.JSONException
import org.json.JSONObject
import java.lang.Exception
import java.math.BigDecimal
import java.util.HashMap

open class TaplyticsKit :
    KitIntegration(),
    AttributeListener,
    KitIntegration.EventListener,
    CommerceListener,
    IdentityListener,
    SessionListener {
    private fun mergeOptions(
        tlOptions: Map<String?, Any?>,
        configuration: MutableMap<String, Any>?,
    ): HashMap<String?, Any?> {
        var tlOptions: Map<String?, Any?>? = tlOptions
        var configuration = configuration
        if (tlOptions == null) {
            tlOptions = HashMap()
        }
        if (configuration == null) {
            configuration = HashMap()
        }
        val merged = HashMap(configuration)
        for ((key, value) in tlOptions) {
            merged[key] = value
        }
        return merged
    }

    public override fun onKitCreate(
        settings: Map<String, String>,
        context: Context,
    ): List<ReportingMessage> {
        if (!started && !delayInitializationUntilSessionStart) {
            started = true
            startTaplytics(settings, context)
        }
        return emptyList()
    }

    private fun createReportingMessages(message: String): List<ReportingMessage> {
        val reportingMessage =
            ReportingMessage(
                this,
                message,
                System.currentTimeMillis(),
                null,
            )
        return listOf(reportingMessage)
    }

    protected open fun startTaplytics(
        settings: Map<String, String>,
        context: Context?,
    ) {
        val apiKey = getAPIKey(settings)
        val options = mergeOptions(tlOptions, getOptionsFromConfiguration(settings))
        options[DELAYED_START] = true
        Taplytics.startTaplytics(context, apiKey, options)
    }

    private fun getAPIKey(settings: Map<String, String>): String? {
        val apiKey = settings[API_KEY]
        require(!KitUtils.isEmpty(apiKey)) { FAILED_TO_INITIALIZE_KIT_MESSAGE }
        return apiKey
    }

    private fun getOptionsFromConfiguration(settings: Map<String, String>): MutableMap<String, Any>? {
        val options = HashMap<String, Any>()
        addAggressiveOption(options, settings)
        return if (options.isEmpty()) null else options
    }

    private fun addAggressiveOption(
        options: MutableMap<String, Any>,
        settings: Map<String, String>,
    ) {
        val agg = settings[AGGRESSIVE].toBoolean()
        options[TAPLYTICS_AGGRESSIVE] = agg
    }

    override fun getName(): String = KIT_NAME

    /**
     * AttributeListener Interface
     */
    override fun setUserAttribute(
        attributeKey: String,
        attributeValue: String?,
    ) {
        try {
            val attr = JSONObject()
            if (attributeValue != null) {
                attr.put(attributeKey, attributeValue)
            } else {
                attr.put(attributeKey, "")
            }
            Taplytics.setUserAttributes(attr)
        } catch (e: JSONException) {
        }
    }

    override fun supportsAttributeLists(): Boolean = false

    override fun setAllUserAttributes(
        attributes: Map<String, String>,
        attributeLists: Map<String, List<String>>,
    ) {
        for ((key, value) in attributes) {
            setUserAttribute(key, value)
        }
    }

    override fun setUserIdentity(
        identityType: IdentityType,
        s: String?,
    ) {
        // no-op
    }

    override fun removeUserIdentity(identityType: IdentityType) {
        setUserIdentity(identityType, null)
    }

    override fun removeUserAttribute(attribute: String) {
        setUserAttribute(attribute, null)
    }

    override fun onIdentifyCompleted(
        mParticleUser: MParticleUser,
        filteredIdentityApiRequest: FilteredIdentityApiRequest,
    ) {
        setUserAttributeFromRequest(filteredIdentityApiRequest)
    }

    /**
     * Identity Listener
     */
    private fun setUserAttributeFromRequest(request: FilteredIdentityApiRequest) {
        val identities = request.userIdentities
        try {
            val attr = JSONObject()
            if (identities[IdentityType.CustomerId] != null) {
                attr.put(USER_ID, identities[IdentityType.CustomerId])
            }
            if (identities[IdentityType.Email] != null) {
                attr.put(EMAIL, identities[IdentityType.Email])
            }
            Taplytics.setUserAttributes(attr)
        } catch (e: JSONException) {
        }
    }

    override fun onLoginCompleted(
        user: MParticleUser,
        request: FilteredIdentityApiRequest,
    ) {
        setUserAttributeFromRequest(request)
    }

    override fun onLogoutCompleted(
        mParticleUser: MParticleUser,
        filteredIdentityApiRequest: FilteredIdentityApiRequest,
    ) {
        Taplytics.resetAppUser {
            // no-op
        }
    }

    override fun onModifyCompleted(
        mParticleUser: MParticleUser,
        request: FilteredIdentityApiRequest,
    ) {
        setUserAttributeFromRequest(request)
    }

    override fun onUserIdentified(mParticleUser: MParticleUser) {
        // no-op
    }

    /**
     * Unsupported methods
     */
    override fun logout(): List<ReportingMessage> = emptyList()

    override fun setUserAttributeList(
        attribute: String,
        attributeValueList: List<String>,
    ) {}

    /**
     * CommerceListener Interface
     */
    override fun logEvent(event: CommerceEvent): List<ReportingMessage> {
        if (!KitUtils.isEmpty(event.productAction) &&
            event.productAction.equals(Product.PURCHASE, true)
        ) {
            val transactionAttributes = event.transactionAttributes ?: return emptyList()
            val id = transactionAttributes.id
            val revenue = transactionAttributes.revenue ?: return emptyList()
            Taplytics.logRevenue(id, revenue)
            return listOf(ReportingMessage.fromEvent(this, event))
        }
        return emptyList()
    }

    /**
     * EventListener Interface
     */
    override fun logEvent(event: MPEvent): List<ReportingMessage>? {
        val eventName = event.eventName
        val metaDataMap = event.customAttributeStrings
        var metaData: JSONObject? = null
        if (metaDataMap != null) {
            metaData = (metaDataMap as Map<*, *>?)?.let { JSONObject(it) }
        }
        Taplytics.logEvent(eventName, null, metaData)
        return listOf(ReportingMessage.fromEvent(this, event))
    }

    override fun logScreen(
        screenName: String,
        screenAttributes: Map<String, String>,
    ): List<ReportingMessage> {
        Taplytics.logEvent(screenName)
        return createReportingMessages(ReportingMessage.MessageType.SCREEN_VIEW)
    }

    /**
     * Unsupported Methods
     */
    override fun logException(
        exception: Exception,
        exceptionAttributes: Map<String, String>,
        message: String,
    ): List<ReportingMessage> = emptyList()

    override fun logError(
        message: String,
        errorAttributes: Map<String, String>,
    ): List<ReportingMessage> = emptyList()

    override fun leaveBreadcrumb(breadcrumb: String): List<ReportingMessage> = emptyList()

    // put all these together
    override fun logLtvIncrease(
        valueIncreased: BigDecimal,
        valueTotal: BigDecimal,
        eventName: String,
        contextInfo: Map<String, String>,
    ): List<ReportingMessage> = emptyList()

    override fun setOptOut(optedOut: Boolean): List<ReportingMessage> {
        Taplytics.hasUserOptedOutTracking(null) { hasOptedOut ->
            if (!hasOptedOut && optedOut) {
                Taplytics.optOutUserTracking(null)
            } else if (hasOptedOut && !optedOut) {
                Taplytics.optInUserTracking(null)
            }
        }
        return createReportingMessages(ReportingMessage.MessageType.OPT_OUT)
    }

    override fun onSessionStart(): List<ReportingMessage> {
        if (!started && delayInitializationUntilSessionStart) {
            started = true
            startTaplytics(configuration.settings, context)
        }
        return emptyList()
    }

    override fun onSessionEnd(): List<ReportingMessage> = emptyList()

    companion object {
        /**
         * Option Keys
         */
        private const val API_KEY = "apiKey"
        private const val AGGRESSIVE = "TaplyticsOptionAggressive"
        private const val TAPLYTICS_AGGRESSIVE = "aggressive"
        private const val USER_ID = "user_id"
        private const val EMAIL = "email"
        private const val DELAYED_START = "delayedStartTaplytics"
        private const val FAILED_TO_INITIALIZE_KIT_MESSAGE = "Failed to initialize Taplytics SDK - an API key is required"
        private const val KIT_NAME = "Taplytics"

        @JvmField
        var delayInitializationUntilSessionStart = false

        @JvmField
        var started = false

        /**
         * tlOptions get and set methods
         */
        var tlOptions = HashMap<String?, Any?>()
    }
}
